isthmus/src/main/java/io/substrait/isthmus/expression/RexExpressionConverter.java

```diff
@@ -3,29 +3,42 @@
 import io.substrait.expression.Expression;
 import io.substrait.expression.FieldReference;
 import io.substrait.isthmus.CallConverter;
+import io.substrait.isthmus.SubstraitRelVisitor;
 import io.substrait.isthmus.TypeConverter;
+import io.substrait.relation.Rel;
 import io.substrait.type.StringTypeVisitor;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.stream.Collectors;
 import org.apache.calcite.rex.*;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
 
 public class RexExpressionConverter implements RexVisitor<Expression> {
   static final org.slf4j.Logger logger =
       org.slf4j.LoggerFactory.getLogger(RexExpressionConverter.class);
 
   private final List<CallConverter> callConverters;
+  private final SubstraitRelVisitor relVisitor;
+
+  public RexExpressionConverter(SubstraitRelVisitor relVisitor, CallConverter... callConverters) {
```

Comment:
Can we use something narrower than SubstraitRelVisitor as the argument here?

Reply:
Here I have to use SubstraitRelVisitor as we need to get the `steps_out` value (which requires a view of the entire RelNode Tree). The RexExpression (only have access to a rex expression tree) does not have that knowledge.

Reply:
Couldn't this information carried down on the traversal (rather than walking back up the tree once you hit the subquery)? Seems like that would reduce the number of times you're traversing the tree (if I understand this correctly). In that case, we'd probably carry something more akin to a map with correlation to distance from current location.

Reply:
The `fieldAccessDepthMap` is built on the initial traversal in `SubstraitRelVisitor.popFieldAccessDepthMap`.  In `RexExpressionConverter`, it use the `SubstraitRelVisitor's map to get `steps_out`.  So, in total, there are 2 times of tree traversal in `RelNode -> substrait Rel` conversion.

Reply:
ok. Let's keep in mind this may be a place to find perf later if needed.